New N2 capability to save and load views.

Kenneth-T-Moore opened the pull request:
### Summary

New N2 capability to save and load views.
- Former "save svg" button is now a button group with 3 buttons: save svg, save view, load view.
- Removed some extra stuff in the icon.css that appeared to be left over from older versions (please check during review)
- Modified icon instructions so we don't lose one of the free icons.
- A hash of the model's data structure is now computed when the modeldata is saved/recorded so that we don't load a view from a different model

### Related Issues

- Resolves #1612

### Backwards incompatibilities

None

### New Dependencies

None

tadkollar commented:
The primary functionality of saving and restoring a view works well!

Some issues:
- the forward/back button history isn't restored after loading - I felt like I wanted to have it there, but it may not matter to others.
- there are two save dialogs that come up... one to just type in a filename and then another with the filesystem browser. But the same happens for `Save to SVG`, which hasn't been updated in awhile so I didn't notice when that began. Ideally we just want the filesystem browser. If this isn't trivial to fix, we should create a separate issue for it.
- if you try to load a file that isn't a state (or even a JSON file), it bombs. Is there a way to check the validity of the file first?
- the toolbar spacing and sizing got wonky. Some of the buttons shrank (orange annotation), the pullout arrows shifted right (purple annotation), and the pullout expansions are no longer flush (red annotation)
- ![pr_screenshot](https://user-images.githubusercontent.com/12797795/108098382-4c517200-7051-11eb-86d6-93e00ff511e2.png)![master_screenshot](https://user-images.githubusercontent.com/12797795/108098374-49ef1800-7051-11eb-9d22-ea4ebcfe4cfc.png)

Kenneth-T-Moore commented:
Thanks, Tad

1. _the forward/back button history isn't restored after loading_

Now that I have the basics working, I could probably add that.  Not sure how much that will impact the file size.

2. _there are two save dialogs that come up..._ 

Yeah, this is kind of clunky.  I'll look into this a little, but maybe you are right about it being a separate issue.

3. if you try to load a file that isn't a state (or even a JSON file), it bombs

Yeah, I should just check that the top level dictionary keys are there, and if they aren't, raise an error.

4. the toolbar spacing and sizing got wonky. Some of the buttons shrank (orange annotation), the pullout arrows shifted right (purple annotation), and the pullout expansions are no longer flush (red annotation)

Herb mentioned that his happened in the past, but didn't know how to fix it. Do you have any ideas? It changed when I created the new Icomoon font .woff file.

tadkollar commented:
> Herb mentioned that his happened in the past, but didn't know how to fix it. Do you have any ideas? It changed when I created the new Icomoon font .woff file.

We ran into this the last time we added a button to the toolbar... the problem is that portion of `icon.css` beneath the docs (i.e. the actual style definitions) is auto-generated by icomoon. It links the font's specific unicode characters to styles, so we copy and paste the css from the zip file that icomoon generates. If you need to make any mods to those styles, they need go into `toolbar.css` or one of the other style sheets.

tadkollar commented on openmdao/visualization/n2_viewer/src/N2Arrow.js:
```diff
@@ -968,4 +980,86 @@ class N2ArrowManager {
+
+        for (var arrowId in pinned.arrows) {
+            let arrow = pinned.arrows[arrowId];
+
```
Using `var` makes a global, so it's usually avoided. Using `let` is OK because it limits the scope to the block, but if you're not going to reassign the variable, `const` is generally the way to go. It creates a constant reference (with block scope), so you can change its properties, but not reassign it. Here, `const arrowId in pinned.arrows` and `const arrow = pinned.arrows[arrowId]` would be good.

Kenneth-T-Moore replied:
Thanks, I will fix these.

Kenneth-T-Moore commented:
_1. The forward/back button history isn't restored after loading_

Based on today's discussion, I modified the loadView code so that it saves the previous view to the back-button stack so the user can restore the view prior to load.

_2. there are two save dialogs that come up..._

I've created issue #1908 for this story.

_3. if you try to load a file that isn't a state (or even a JSON file), it bombs_

Fixed this. Now an alert box is generated to tell the user the file isn't a view.

_4. the toolbar spacing and sizing got wonky. Some of the buttons shrank (orange annotation), the pullout arrows shifted right (purple annotation), and the pullout expansions are no longer flush (red annotation)_

I didn't have any luck with this, and will hold off until your refactor of the icon system to remove the icomoon processing. Hopefully then it will be fully repeatable.

hschilling commented:
- Should the view also save the setting on what to show in the solver section?
- This might be a bug. I saved a view. Did a refresh of the page. Loaded in the view. Then did something to change the view. Hit the move back in history button. So far so good. But then hit the go forward button and nothing happens.
- Probably nothing you can do about this one but when you save a view twice in a row with the same name, the OS, at least on a Mac, saves it with a slightly modified name, which is good. But the you don't get the usual warning about overwriting and keeping both... It behaves just like a browser download.

Kenneth-T-Moore commented:
Hey, Herb. Thanks for checking it out.

- The view does save the solver setting. Seems to work for linear/nonlinear select as well as for just turning off the display of solvers.  I haven't been able to reproduce a case where it doesn't work.  I suppose there's also an argument that could be made that maybe I shouldn't save the solver state in the view, but I am not sure.
- I couldn't reproduce. One thing that can cause confusion is that forward/backward history only saves the zoomed node and collapsed state, so if your view has some highlighted some connections, those stay on when you go backwards in history.
- This is a good point. We have a story (#1908) up to completely redo the file interaction to use a browser (instead of having two dialog boxes), so I think (hope) it will be fixed with this story.

hschilling commented:
Now I can't make those two errors show up either so never mind! I was especially careful to document and do over and over the second one. Not sure why that isn't doing it now. 

One final comment. Since these are called "views" should the default file extension be ".view" instead of ".state" ?

swryan commented:
> One final comment. Since these are called "views" should the default file extension be ".view" instead of ".state" ?

or more specifically ".n2view"?

Kenneth-T-Moore commented:
Good points. I will make it ".n2view".
